src/visitors/cssNamespace.js

```diff
@@ -1,122 +1,144 @@
-        hasNamespaceSelfReference ||
-        startsWithSelfReference(quasis[0].value.raw);
-
-      const first = t.templateElement(
-        {
-          cooked: `\n${wrapCssDefinition(
-            cssNamespace,
-            quasis[0].value.cooked,
-            isCookedSelfReferenced,
-            STRIP_END_CURLY_BRACKETS
-          )}`,
-          raw: `\n${wrapCssDefinition(
-            cssNamespace,
-            quasis[0].value.raw,
-            isRawSelfReferenced,
-            STRIP_END_CURLY_BRACKETS
-          )}`
-        },
-        quasis[0].tail
-      );
-
-      const last = t.templateElement(
-        {
-          cooked: `${quasis[quasis.length - 1].value
-            .cooked}${isCookedSelfReferenced ? '}' : '}}'}\n`,
-          raw: `${quasis[quasis.length - 1].value.raw}${isRawSelfReferenced
-            ? '}'
-            : '}}'}\n`
-        },
-        quasis[quasis.length - 1].tail
-      );
-      values = [first, ...quasis.slice(1, quasis.length - 1), last];
-    }
-
-    path.replaceWith(
-      t.taggedTemplateExpression(callee, t.templateLiteral(values, expressions))
-    );
+  // Ignore nodes generated by this visitor, to prevent infinite loops
+  if (replacementNodes.has(node)) return;
+
+  // Ignore templates tagged with anything other than `styled(x)`
+  // Inspired by https://github.com/TrevorBurnham/babel-plugin-namespace-styled-components
+  if (!isStyled(t)(tag, state)) return;
+  if (isKeyframesHelper(t)(tag, state)) return;
+  if (isPureHelper(t)(tag, state)) return;
+  if (isHelper(t)(tag, state)) return;
+  if (isCSSHelper(t)(tag, state)) return;
+  if (isInjectGlobalHelper(t)(tag, state)) return;
+  if (tag.property && tag.property.name === 'keyframes') return; // Maintain backward compatibility with styled.keyframes
+
+  // Convert the tagged template to a string, with ${} expressions replaced with placeholders
+  const originalStyleString = quasis
+    .map((quasi, i) => {
+      const rawValue = quasi.value.raw;
+      const nextQuasi = quasis[i + 1];
+      const rawValueWithoutWhiteSpace = rawValue.replace(/[\n\r\s]/g, '');
+
+      // When there is no associated expression, we can return the plain string
+      if (!expressions[i]) {
+        return rawValue;
+      }
+
+      // In cases where the expression would result in invalid css, we need to add
+      // a fake value.
+      // For example,
+      // .button {
+      //   ${someVariableWithValidCss};
+      // }
+      // Without the following logic that would return as:
+      // .button {
+      //   EXPRESSION;
+      // }
+      // Which even the safe parser doesn't know how to handle.
+      // So we add a fake value to make it valid css.
+      // .button {
+      //   EXPRESSION: fakevalue;
+      // }
+      if (
+        (rawValueWithoutWhiteSpace === '' ||
+          rawValueWithoutWhiteSpace.endsWith('{') ||
+          rawValueWithoutWhiteSpace.endsWith('}') ||
+          rawValueWithoutWhiteSpace.endsWith(';')) &&
```

Comment:
I don't love this, but I haven't found a better way to detect when an expression appears on its own line and causes invalid css.

Reply:
I ended up hacking the parser to solve this: https://github.com/TrevorBurnham/babel-plugin-namespace-styled-components/pull/2

Reply:
Awesome. Thanks for the heads up. That might solve another issue that was opened as well.